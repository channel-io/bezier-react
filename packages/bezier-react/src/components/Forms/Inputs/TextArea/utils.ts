import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
