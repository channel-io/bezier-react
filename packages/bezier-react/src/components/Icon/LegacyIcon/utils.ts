/* Internal Dependencies */
// eslint-disable-next-line no-restricted-imports
import icons, { type IconName } from '../generated'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { Icon, AllIcon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-react'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(<svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
