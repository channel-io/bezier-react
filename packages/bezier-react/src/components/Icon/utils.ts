import { type SVGProps } from 'react'

const BEZIER_ICON_SYMBOL = Symbol('bezier.icon')

export type IconSource = (props: SVGProps<SVGSVGElement>) => JSX.Element

export type BezierIcon = IconSource & { [BEZIER_ICON_SYMBOL]: true }

export function isBezierIcon(arg: unknown): arg is BezierIcon {
  return (
    typeof arg === 'function' &&
    arg[BEZIER_ICON_SYMBOL] === true
  )
}

export function createBezierIcon(source: IconSource): BezierIcon {
  source[BEZIER_ICON_SYMBOL] = true
  return source as BezierIcon
}
