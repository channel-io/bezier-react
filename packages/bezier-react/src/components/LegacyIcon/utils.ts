import {
  type IconName,
  icons,
} from '@channel.io/bezier-icons'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { AllIcon } from '@channel.io/bezier-icons'
 * import { Icon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-icons'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(() => <svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
