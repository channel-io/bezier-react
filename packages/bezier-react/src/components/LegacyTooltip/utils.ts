import {
  type GetReplacement,
  type GetTooltipStyle,
  LegacyTooltipPosition,
} from './LegacyTooltip.types'

export function getReplacement({
  tooltip,
  keepInContainer,
  placement,
  rootElement,
}: GetReplacement): LegacyTooltipPosition {
  if (!keepInContainer) {
    return placement
  }

  const {
    width: tooltipWidth,
    height: tooltipHeight,
    top: tooltipTop,
    left: tooltipLeft,
  } = tooltip.getBoundingClientRect()

  const {
    width: rootWidth,
    height: rootHeight,
    top: rootTop,
    left: rootLeft,
  } = rootElement.getBoundingClientRect()

  const isOverTop = tooltipTop < rootTop
  const isOverBottom = tooltipTop + tooltipHeight > rootTop + rootHeight
  const isOverLeft = tooltipLeft < rootLeft
  const isOverRight = tooltipLeft + tooltipWidth > rootLeft + rootWidth

  if (isOverTop) {
    if (isOverLeft) {
      return LegacyTooltipPosition.BottomLeft
    }
    if (isOverRight) {
      return LegacyTooltipPosition.BottomRight
    }
    return LegacyTooltipPosition.BottomCenter
  }

  if (isOverBottom) {
    if (isOverLeft) {
      return LegacyTooltipPosition.TopLeft
    }
    if (isOverRight) {
      return LegacyTooltipPosition.TopRight
    }
    return LegacyTooltipPosition.TopCenter
  }

  if (isOverLeft) {
    if (isOverTop) {
      return LegacyTooltipPosition.RightTop
    }
    if (isOverBottom) {
      return LegacyTooltipPosition.RightBottom
    }
    return LegacyTooltipPosition.RightCenter
  }

  if (isOverRight) {
    if (isOverTop) {
      return LegacyTooltipPosition.LeftTop
    }
    if (isOverBottom) {
      return LegacyTooltipPosition.LeftBottom
    }
    return LegacyTooltipPosition.LeftCenter
  }

  return placement
}

export function getTooltipStyle({
  tooltipContainer,
  placement,
  offset,
  allowHover,
}: GetTooltipStyle) {
  const {
    top: containerTop,
    left: containerLeft,
    width: containerWidth,
    height: containerHeight,
  } = tooltipContainer.getBoundingClientRect()

  let top = containerTop
  let left = containerLeft
  let translateX = 0
  let translateY = 0

  let offsetTop = 0
  let offsetRight = 0
  let offsetBottom = 0
  let offsetLeft = 0
  let offsetStyle = {}

  switch (placement) {
    case LegacyTooltipPosition.TopCenter:
    case LegacyTooltipPosition.TopLeft:
    case LegacyTooltipPosition.TopRight:
      translateY = -100
      offsetBottom = offset
      break
    case LegacyTooltipPosition.RightCenter:
    case LegacyTooltipPosition.RightTop:
    case LegacyTooltipPosition.RightBottom:
      left += containerWidth
      offsetLeft = offset
      break
    case LegacyTooltipPosition.BottomCenter:
    case LegacyTooltipPosition.BottomLeft:
    case LegacyTooltipPosition.BottomRight:
      top += containerHeight
      offsetTop = offset
      break
    case LegacyTooltipPosition.LeftCenter:
    case LegacyTooltipPosition.LeftTop:
    case LegacyTooltipPosition.LeftBottom:
      translateX = -100
      offsetRight = offset
      break
  }

  switch (placement) {
    case LegacyTooltipPosition.TopCenter:
    case LegacyTooltipPosition.BottomCenter:
      translateX = -50
      left += (containerWidth / 2)
      break
    case LegacyTooltipPosition.TopRight:
    case LegacyTooltipPosition.BottomRight:
      translateX = -100
      left += containerWidth
      break
    case LegacyTooltipPosition.RightCenter:
    case LegacyTooltipPosition.LeftCenter:
      translateY = -50
      top += (containerHeight / 2)
      break
    case LegacyTooltipPosition.RightBottom:
    case LegacyTooltipPosition.LeftBottom:
      translateY = -100
      top += containerHeight
      break
  }

  if (allowHover) {
    offsetStyle = {
      paddingTop: offsetTop,
      paddingRight: offsetRight,
      paddingBottom: offsetBottom,
      paddingLeft: offsetLeft,
    }
  } else {
    top += (offsetTop - offsetBottom)
    left += (offsetLeft - offsetRight)
  }

  return {
    position: 'fixed',
    top,
    left,
    transform: `translate(${translateX}%, ${translateY}%)`,
    ...offsetStyle,
  }
}
