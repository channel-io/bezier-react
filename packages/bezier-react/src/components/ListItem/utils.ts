import { css } from '~/src/foundation'

import { ListItemSize } from './ListItem.types'

export function getStyleOfSize(size?: ListItemSize) {
  switch (size) {
    case ListItemSize.S:
      return css`
        padding: 4px 6px;
        ${({ foundation }) => foundation?.rounding.round6};
      `
    case ListItemSize.L:
      return css`
        padding: 8px 6px;
        ${({ foundation }) => foundation?.rounding.round8};
      `
    case ListItemSize.XL:
      return css`
        padding: 10px 6px;
        ${({ foundation }) => foundation?.rounding.round12};
      `
    case ListItemSize.M:
    default:
      return css`
        padding: 6px;
        ${({ foundation }) => foundation?.rounding.round6};
      `
  }
}
