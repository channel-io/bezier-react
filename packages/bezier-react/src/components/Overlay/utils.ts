/* Internal dependencies */
import { css } from '~/src/foundation'
import {
  OverlayPosition,
  type ContainerRectAttr,
  type TargetRectAttr,
} from './Overlay.types'

const TOP_POSITION_OFFSET = 10

interface GetOverlayPositionArgs {
  containerRect: ContainerRectAttr
  targetRect: TargetRectAttr
  show: boolean
}

export function getOverlayPosition({
  containerRect,
  targetRect,
  show,
}: GetOverlayPositionArgs) {
  const { containerTop, containerLeft, scrollTop, scrollLeft } = containerRect
  const { targetTop, targetLeft, clientTop, clientLeft } = targetRect

  const top = targetTop - clientTop - containerTop + scrollTop
  const left = targetLeft - clientLeft - containerLeft + scrollLeft

  return {
    top: show ? top : top - TOP_POSITION_OFFSET,
    left,
  }
}

interface GetOverlayTranslatationArgs {
  containerRect: ContainerRectAttr
  targetRect: TargetRectAttr | null
  overlay: HTMLElement | null
  position: OverlayPosition
  marginX: number
  marginY: number
  keepInContainer: boolean
}

export function getOverlayTranslation({
  containerRect,
  targetRect,
  overlay,
  position,
  marginX,
  marginY,
  keepInContainer,
}: GetOverlayTranslatationArgs) {
  if (containerRect && targetRect && overlay) {
    const {
      containerWidth,
      containerHeight,
      containerTop,
      containerLeft,
    } = containerRect
    const { targetWidth, targetHeight, targetTop, targetLeft } = targetRect
    const { width: overlayWidth, height: overlayHeight } = overlay.getBoundingClientRect()

    let translateX = 0
    let translateY = 0

    // pre position
    switch (position) {
      case OverlayPosition.TopCenter:
      case OverlayPosition.TopLeft:
      case OverlayPosition.TopRight:
        translateY -= (overlayHeight + marginY)
        translateX += marginX
        break
      case OverlayPosition.RightCenter:
      case OverlayPosition.RightTop:
      case OverlayPosition.RightBottom:
        translateX += (targetWidth + marginX)
        translateY += marginY
        break
      case OverlayPosition.BottomCenter:
      case OverlayPosition.BottomLeft:
      case OverlayPosition.BottomRight:
        translateY += (targetHeight + marginY)
        translateX += marginX
        break
      case OverlayPosition.LeftCenter:
      case OverlayPosition.LeftTop:
      case OverlayPosition.LeftBottom:
        translateX -= (overlayWidth + marginX)
        translateY += marginY
        break
    }

    switch (position) {
      // post position
      case OverlayPosition.TopCenter:
      case OverlayPosition.BottomCenter:
        translateX -= ((overlayWidth / 2) - (targetWidth / 2))
        break
      case OverlayPosition.TopRight:
      case OverlayPosition.BottomRight:
        translateX -= (overlayWidth - targetWidth)
        break
      case OverlayPosition.RightCenter:
      case OverlayPosition.LeftCenter:
        translateY -= ((overlayHeight / 2) - (targetHeight / 2))
        break
      case OverlayPosition.RightBottom:
      case OverlayPosition.LeftBottom:
        translateY -= (overlayHeight - targetHeight)
        break

      // inner position
      case OverlayPosition.InnerLeftTop:
        translateX = marginX
        translateY = marginY
        break
      case OverlayPosition.InnerLeftBottom:
        translateX = marginX
        translateY = targetHeight - overlayHeight + marginY
        break
      case OverlayPosition.InnerRightTop:
        translateX = targetWidth - overlayWidth + marginX
        translateY = marginY
        break
      case OverlayPosition.InnerRightBottom:
        translateX = targetWidth - overlayWidth + marginX
        translateY = targetHeight - overlayHeight + marginY
        break
    }

    if (keepInContainer) {
      const isOverTop = targetTop + translateY < containerTop
      const isOverBottom = targetTop + translateY + overlayHeight > containerTop + containerHeight
      const isOverLeft = targetLeft + translateX < containerLeft
      const isOverRight = targetLeft + translateX + overlayWidth > containerLeft + containerWidth

      const topSpace = targetTop - containerTop
      const bottomSpace = (containerTop + containerHeight) - (targetTop + targetHeight)
      const leftSpace = targetLeft - containerLeft
      const rightSpace = (containerLeft + containerWidth) - (targetLeft + targetWidth)

      if (isOverTop) {
        translateY = topSpace > bottomSpace
          ? translateY
          : targetHeight - translateY - overlayHeight
      } else if (isOverBottom) {
        translateY = bottomSpace > topSpace
          ? translateY
          : targetHeight - translateY - overlayHeight
      }
      if (isOverLeft) {
        translateX = leftSpace > rightSpace
          ? translateX
          : targetWidth - translateX - overlayWidth
      } else if (isOverRight) {
        translateX = rightSpace > leftSpace
          ? translateX
          : targetWidth - translateX - overlayWidth
      }
    }

    return {
      translateX: Math.round(translateX),
      translateY: Math.round(translateY),
    }
  }
  return {
    translateX: 0,
    translateY: 0,
  }
}

interface GetOverlayStyleArgs {
  containerRect: ContainerRectAttr
  targetRect: TargetRectAttr | null
  overlay: HTMLElement | null
  position: OverlayPosition
  marginX: number
  marginY: number
  keepInContainer: boolean
  show: boolean
}

export function getOverlayStyle({
  containerRect,
  targetRect,
  overlay,
  position,
  marginX,
  marginY,
  keepInContainer,
  show,
}: GetOverlayStyleArgs) {
  if (containerRect && targetRect) {
    const { top, left } = getOverlayPosition({ containerRect, targetRect, show })
    const { translateX, translateY } = getOverlayTranslation({
      containerRect,
      targetRect,
      overlay,
      position,
      marginX,
      marginY,
      keepInContainer,
    })

    return css`
      top: ${top}px;
      left: ${left}px;
      transform: translateX(${translateX}px) translateY(${translateY}px);
    `
  }
  return css``
}
