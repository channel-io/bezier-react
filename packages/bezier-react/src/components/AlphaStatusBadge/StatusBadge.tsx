import React, { type CSSProperties, forwardRef } from 'react'

import { MoonFilledIcon } from '@channel.io/bezier-icons'
import classNames from 'classnames'

import { type SemanticColor } from '~/src/types/tokens'
import { cssVar } from '~/src/utils/style'

import { Icon } from '~/src/components/Icon'

import { type StatusBadgeProps } from './StatusBadge.types'

import styles from './StatusBadge.module.scss'

// TODO: Change to use AlphaIcon and alpha color tokens

/**
 * `StatusBadge` is a component to indicate user status.
 * @example
 *
 * ```tsx
 * <StatusBadge
 *   size="m"
 *   online
 *   doNotDisturb
 * />
 * ```
 */
export const StatusBadge = forwardRef<HTMLDivElement, StatusBadgeProps>(
  function StatusBadge(
    {
      size = 'm',
      online = false,
      doNotDisturb = false,
      style,
      className,
      ...rest
    },
    forwardedRef
  ) {
    const iconColor: SemanticColor = online
      ? 'bgtxt-green-normal'
      : doNotDisturb
        ? 'bgtxt-yellow-normal'
        : 'bg-black-dark'
    const backgroundColor: SemanticColor = doNotDisturb
      ? 'bg-white-high'
      : iconColor

    return (
      <div
        ref={forwardedRef}
        style={
          {
            '--b-status-bg-color': cssVar(backgroundColor),
            ...style,
          } as CSSProperties
        }
        className={classNames(
          styles.StatusBadge,
          styles[`size-${size}`],
          className
        )}
        {...rest}
      >
        {doNotDisturb && (
          <Icon
            source={MoonFilledIcon}
            size={size === 'm' ? 'xxxs' : 'xs'}
            color={iconColor}
            className={styles.Icon}
          />
        )}
      </div>
    )
  }
)
