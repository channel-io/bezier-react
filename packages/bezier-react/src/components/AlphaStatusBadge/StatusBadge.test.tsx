import React from 'react'

import { render } from '~/src/utils/test'

import { StatusBadge } from './StatusBadge'
import { type StatusBadgeProps } from './StatusBadge.types'

describe('StatusBadge', () => {
  let props: StatusBadgeProps

  beforeEach(() => {
    props = {}
  })

  afterAll(() => {
    jest.resetAllMocks()
  })

  const renderComponent = (otherProps?: StatusBadgeProps) =>
    render(
      <StatusBadge
        {...props}
        {...otherProps}
      />
    )

  it('should match snapshot', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })

  it('medium status badge', () => {
    const { container } = renderComponent({ size: 'm' })
    expect(container).toMatchSnapshot()
  })
  it('large status badge', () => {
    const { container } = renderComponent({ size: 'l' })
    expect(container).toMatchSnapshot()
  })

  it('online status', () => {
    const { container } = renderComponent({ online: true })
    expect(container).toMatchSnapshot()
  })
  it('offline status', () => {
    const { container } = renderComponent({ online: false })
    expect(container).toMatchSnapshot()
  })

  it('do not disturb (online) status', () => {
    const { container } = renderComponent({ doNotDisturb: true, online: true })
    expect(container).toMatchSnapshot()
  })
  it('do not disturb (offline) status', () => {
    const { container } = renderComponent({ doNotDisturb: true, online: false })
    expect(container).toMatchSnapshot()
  })
})
