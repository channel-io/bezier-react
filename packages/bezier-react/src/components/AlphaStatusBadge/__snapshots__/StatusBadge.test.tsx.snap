// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusBadge do not disturb (offline) status 1`] = `
<div>
  <div
    class="Status size-m"
    style="--b-status-bg-color: var(--bg-white-high);"
  >
    <svg
      class="Icon size-xxxs margin Icon"
      data-testid="bezier-icon"
      fill="none"
      height="1em"
      style="--b-icon-color: var(--bgtxt-yellow-normal);"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20.64 17.037c.458-.782-.454-1.568-1.33-1.336A9 9 0 0 1 8.3 4.69c.231-.877-.555-1.789-1.337-1.332A10 10 0 0 0 2 12c0 5.523 4.477 10 10 10a10 10 0 0 0 8.64-4.963"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;

exports[`StatusBadge do not disturb (online) status 1`] = `
<div>
  <div
    class="Status size-m"
    style="--b-status-bg-color: var(--bg-white-high);"
  >
    <svg
      class="Icon size-xxxs margin Icon"
      data-testid="bezier-icon"
      fill="none"
      height="1em"
      style="--b-icon-color: var(--bgtxt-green-normal);"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20.64 17.037c.458-.782-.454-1.568-1.33-1.336A9 9 0 0 1 8.3 4.69c.231-.877-.555-1.789-1.337-1.332A10 10 0 0 0 2 12c0 5.523 4.477 10 10 10a10 10 0 0 0 8.64-4.963"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;

exports[`StatusBadge large status badge 1`] = `
<div>
  <div
    class="Status size-l"
    style="--b-status-bg-color: var(--bg-black-dark);"
  />
</div>
`;

exports[`StatusBadge medium status badge 1`] = `
<div>
  <div
    class="Status size-m"
    style="--b-status-bg-color: var(--bg-black-dark);"
  />
</div>
`;

exports[`StatusBadge offline status 1`] = `
<div>
  <div
    class="Status size-m"
    style="--b-status-bg-color: var(--bg-black-dark);"
  />
</div>
`;

exports[`StatusBadge online status 1`] = `
<div>
  <div
    class="Status size-m"
    style="--b-status-bg-color: var(--bgtxt-green-normal);"
  />
</div>
`;

exports[`StatusBadge should match snapshot 1`] = `
<div>
  <div
    class="Status size-m"
    style="--b-status-bg-color: var(--bg-black-dark);"
  />
</div>
`;
